### ISSUE TITLE
Caddy & Dokku on the same server

### ISSUE THREAD

**Issue by pavlik** [mholt/caddy#375]
How to make friends Caddy & Dokku on the same server?
Sometimes ago I installed Dokku to host my golang webapp. 
Today I want to move all my hugo static sites to this server too.
This is my Caddyfile
```
http://grapefilms.pro, http://www.grapefilms.pro {
        root /var/www/grapefilms.pro/public/
        gzip
        log access.log

        errors {
                log error.log
                404 404.html # Not Found
        }

#       git {
#               repo  https://github.com/pavlik/grapefilms.pro-on-hugo
#               path  ../
#               hook /webhook secret
#               then  hugo --destination=public
#       }
}

http://village-geek.ru, http://www.village-geek.ru {
        root /var/www/village-geek.ru/public/
        gzip
        log access.log

        errors {
                log error.log
                404 404.html # Not Found
        }

#       git {
#               repo  https://github.com/pavlik/village-geek.ru
#               path  ../
#               hook /webhook secret
#               then  hugo hugo --theme=hugo-theme-blueberry-detox --destination=public
#       }
}
```
If I start Caddy then I receive message *"2015/12/01 14:26:33 listen tcp :80: bind: address already in use"*

P.S. I commented git section because if it is uncommented I receive error: *Caddyfile:11 - Parse error: Unknown directive 'git'*

**Comment by pavlik** [comment]
Maybe if I installed Caddy then Dokku is no longer needed & I can config automatic pull & build via 'git' add-on? i.e.:
```
http://mimozaflowers.ru, http://www.mimozaflowers.ru {
        root /var/www/mimozaflowers.ru/
        gzip
        log access.log

        errors {
                log error.log
                404 404.html # Not Found
        }

      git {
               repo  https://github.com/pavlik/mimozaflowers
               hook /webhook secret
               then  go build
       }

some command that start my app????
```

**Comment by abiosoft** [comment]
Yeah, you can. 

You will need to download caddy with git add-on via the caddy download page.

Right now, git only supports short running commands. I am assuming after `go build`, you would want to execute the resulting binary. It actually waits for the command to finish executing before continuing.

Now I am seeing a usecase for long running commands and I am considering adding the feature to git.

**Comment by pavlik** [comment]
@abiosoft thanks! How to download it via terminal?

**Comment by abiosoft** [comment]
wget `http://caddyserver.com/download/build?os=linux&arch=amd64&features=git`

**Comment by pavlik** [comment]
@abiosoft 
```
Resolving caddyserver.com (caddyserver.com)... 162.243.142.63
Connecting to caddyserver.com (caddyserver.com)|162.243.142.63|:80... connected.
HTTP request sent, awaiting response... 400 Bad Request
2015-12-01 17:25:03 ERROR 400: Bad Request.
```

**Comment by abiosoft** [comment]
Sorry, put the URL in quotes.

```
wget "http://caddyserver.com/download/build?os=linux&arch=amd64&features=git" -O "caddy_linux_amd64.tar.gz"
```

**Comment by mholt** [comment]
That "address already in use" error is concerning; similar report in #373 but I haven't been able to reproduce it. Are you sure no other process is already using port 80?

Also, what is the hostname of the machine you are running it on?

**Comment by mholt** [comment]
Along with those questions, I'd also like you to try something for me, if you could; remove hosts (addresses) from the Caddyfile one-by-one and try running Caddy again, until the "address already in use" error goes away. I'm interested which addresses in the Caddyfile combine to cause that error -- and whether it be one server block or multiple. Because I still can't make it happen here...

**Comment by pavlik** [comment]
@mholt 
While I waited for the answer, I did some experiments.
1. Removed dokku
2. Removed the nginx
Now, this error does not appear. But if I run the caddy without param --port = 80 it assigns for my sites port=2015

@abiosoft That will be amazing if git addon can execute long running commands!

**Comment by mholt** [comment]
@pavlik Good, that means that nginx was probably using the port, so naturally Caddy couldn't bind to it.

Using the `-port` flag only specifies the _default_ port, it won't overwrite what's in a Caddyfile.

**Comment by pavlik** [comment]
@abiosoft why I can't execute go command in git add-on? (in terminal go is executed & present in $PATH)
```
015/12/01 21:47:40 https://github.com/pavlik/mimozaflowers.git pulled.
2015/12/01 21:47:40 exec: "go": executable file not found in $PATH
```

@mholt What this message mean & if i set recommended value why it resets after server reboot?
`Warning: File descriptor limit 1024 is too low for production sites. At least 4096 is recommended`

**Comment by abiosoft** [comment]
It means the session executing the command (the one caddy is started in) does not have `go` in PATH.

Can you confirm `go` is accessible before starting caddy.
As a fall back option, you can put absolute path.

**Comment by pavlik** [comment]
@abiosoft yes `go` is accessible before starting caddy.
& if I use absolute path `go` is accessible.

**Comment by mholt** [comment]
So is it working for you now? i.e. was your question about Dokku answered?

**Comment by mholt** [comment]
Ah, now I see abiosoft/caddy-git#13 - closing in favor of that issue.
